File: puzzles-games/MaxRainWater.java

public class MaxRainWater {
    public int trap(int[] height) {
        if (height == null || height.length <= 1) {
            return 0;
        }
        int water = 0;
        int left = 0;
        int right = height.length - 1;
        int highLeft = 0;
        int highRight = 0;
        while (left < right) {
            if (height[left] > height[right]) {
                highRight = Math.max(highRight, height[right]);
                water += (highRight - height[right]);
                right--;
            } else {
                highLeft = Math.max(highLeft, height[left]);
                water += (highLeft - height[left]);
                left++;
            }
        }
        return water;
    }
}

File: puzzles-games/MaxRainWater.cpp

class MaxRainWater {
public:
    int trap(vector<int>& height) {
        int maxWater = 0;
        int left = 0;
        int right = height.size() - 1;
        int highestLeft = 0;
        int highestRight = 0;
        while (left < right) {
            if (height[left] > height[right]) {
                highestRight = max(highestRight, height[right]);
                maxWater += (highestRight - height[right]);
                right--;
            } else {
                highestLeft = max(highestLeft, height[left]);
                maxWater += (highestLeft - height[left]);
                left++;
            }
        }
        return maxWater;
    }
};
